libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1160,6 +1184,39 @@ internal class MapRouteLine(
                 it.addStringProperty(WAYPOINT_PROPERTY_KEY, propValue)
             }
         }
+
+        fun findDistanceOfPointAlongLine(line: LineString, point: Point): Double {
```

Comment:
The `point` argument should be transformed with the `TurfMisc#nearestPointOnLine` before doing the rest of the calculations. Also, we should abort recalculating the vanishing point and leave it as is if the distance to the point on the line is bigger than some threshold which means the precision of map-matching is off. We could take pixel density and size of the puck into this threshold equation if possible:
```kotlin
        fun findDistanceOfPointAlongLine(line: LineString, point: Point): Double? {
          val pointOnLine = TurfMisc.nearestPointOnLine(line.coordinates(), point)
          if (TurfMeasurement.distance(point, pointOnLine) > SOME_THRESHOLD) {
            return null
          }
          ...
          // continue calculations with pointOnLine
```

Reply:
Regarding performance. The longest I've seen the calculation take (on my phone) is 6 ms.

Reply:
Sounds pretty significant if we're doing that on each frame. That'd be more than a third of the time we've got (~16.6ms to maintain 60FPS rendering). If we cannot get it down with other optimizations, we might need to offload to a worker thread, even considering the thread switching penalty and being a frame behind.

Reply:
I don't think TurfMisc.nearestPointOnLine(line.coordinates(), point) tells me what I need to know. What I have to determine is the distance traveled. Knowing that pointOnLine is the closest route point doesn't tell me the distance traveled to get there. It also doesn't tell me if the point parameter is before pointOnLine or after pointOnLine along the route which determines if I need to add or subtract some distance in order to get an accurate value.  The route isn't a straight line so I have to loop through it adding up the distances between points.

Reply:
That was the longes time that it took. Most of the time it was 1-3 ms and rarely up to 6ms but I'll look for optimizations anyway.  I wanted to get a first pass working to see what it looked like.

Reply:
> What I have to determine is the distance traveled.

That's right, but the distance should be measured on the line.

![Screenshot from 2020-08-10 19-28-50](https://user-images.githubusercontent.com/16925074/89812009-c4d5f400-db3f-11ea-84a3-44223c8e44ea.png)

Sorry for the terrible drawing.

The current logic is:
```
distance traveled = x1 + x2 + y1
```
and the vanishing point ends up in front of the puck. The more precise logic is:
```
distance traveled = x1 + x2 + x3
```

That's what the `nearestPointOnLine` gives us. Finding out the position of the point remains the same but we're using the point on the line, not where the puck is.

Since our desired point is located with the `distance` calculation, it'd be great to play around with edge cases like U-turns where the puck might be rendered in the center of the loop and make the vanishing point jump back and forth. Granted most of these use-cases are handled by map-matching (when it's available) which places us in the correct place on the geometry, we should still account for them as much as we can.

Reply:
ok, I misunderstood what nearestPointOnLine did. I thought it looked at the coordinates in the line string and chose the closest one.